File: swift/Sources/AetherNetBitTorrent/Extensions.swift
// SPDX-License-Identifier: MIT

import Foundation

// ── BEP-10 extension protocol ────────────────────────────────────────────────

/// The peer-wire message id for extended messages (BEP-10).
public let extendedMessageID: UInt8 = 20

/// The extended sub-message id of the handshake.
public let extensionHandshakeID: UInt8 = 0

/// Builds an extended message payload: [subID][body]. This is the payload of a
/// peer-wire Extended (id 20) message.
public func wrapExtended(_ subID: UInt8, _ body: [UInt8]) -> [UInt8] {
    var out = [UInt8](repeating: 0, count: 1 + body.count)
    out[0] = subID
    for i in 0..<body.count { out[1 + i] = body[i] }
    return out
}

/// Splits an extended payload into its sub-message id and body.
public func splitExtended(_ payload: [UInt8]) throws -> (UInt8, [UInt8]) {
    if payload.isEmpty {
        throw PeerWireError.invalid("empty extended payload")
    }
    return (payload[0], Array(payload[1...]))
}

/// Builds a BEP-10 handshake advertising supported extensions (name → local
/// sub-message id) and optionally the metadata size.
public func buildExtensionHandshake(_ supported: [String: Int], metadataSize: Int) -> [UInt8] {
    let m = BDict()
    for (name, id) in supported {
        try! m.add(name, .int(Int64(id)))  // Dictionary keys are unique → cannot throw
    }
    let d = BDict()
    try! d.add("m", .dict(m))
    if metadataSize > 0 {
        try! d.add("metadata_size", .int(Int64(metadataSize)))
    }
    return wrapExtended(extensionHandshakeID, bencodeEncode(.dict(d)))
}

/// A parsed BEP-10 handshake.
public struct ExtensionHandshake {
    public var supported: [String: Int]
    public var metadataSize: Int

    /// The peer's ut_metadata sub-message id, or 0 if unsupported.
    public var metadataMessageID: Int { supported["ut_metadata"] ?? 0 }

    /// The peer's ut_pex sub-message id, or 0 if unsupported.
    public var pexMessageID: Int { supported["ut_pex"] ?? 0 }
}

/// Parses a BEP-10 handshake body (the bencode dict after the sub-id).
public func parseExtensionHandshake(_ body: [UInt8]) throws -> ExtensionHandshake {
    var h = ExtensionHandshake(supported: [:], metadataSize: 0)
    let v = try bencodeDecode(body)
    let d = try v.dictValue()
    if let mVal = d.get("m"), let md = try? mVal.dictValue() {
        for name in md.keys {
            if let idVal = md.get(name), let id = try? idVal.intValue() {
                h.supported[name] = Int(id)
            }
        }
    }
    if let sizeVal = d.get("metadata_size"), let n = try? sizeVal.intValue() {
        h.metadataSize = Int(n)
    }
    return h
}

// ── BEP-9 ut_metadata ────────────────────────────────────────────────────────

/// A ut_metadata message type.
public enum MetadataMessageType: Int {
    case request = 0
    case data = 1
    case reject = 2
}

/// The ut_metadata piece size (16 KiB).
public let metadataPieceSize = 16384

/// Builds a ut_metadata request for a piece.
public func buildMetadataRequest(_ piece: Int) -> [UInt8] {
    let d = BDict()
    try! d.add("msg_type", .int(Int64(MetadataMessageType.request.rawValue)))
    try! d.add("piece", .int(Int64(piece)))
    return bencodeEncode(.dict(d))
}

/// Builds a ut_metadata data message (bencode header + raw piece bytes).
public func buildMetadataData(_ piece: Int, _ totalSize: Int, _ data: [UInt8]) -> [UInt8] {
    let d = BDict()
    try! d.add("msg_type", .int(Int64(MetadataMessageType.data.rawValue)))
    try! d.add("piece", .int(Int64(piece)))
    try! d.add("total_size", .int(Int64(totalSize)))
    var out = bencodeEncode(.dict(d))
    out.append(contentsOf: data)
    return out
}

/// Builds a ut_metadata reject message.
public func buildMetadataReject(_ piece: Int) -> [UInt8] {
    let d = BDict()
    try! d.add("msg_type", .int(Int64(MetadataMessageType.reject.rawValue)))
    try! d.add("piece", .int(Int64(piece)))
    return bencodeEncode(.dict(d))
}

/// A parsed ut_metadata message.
public struct MetadataMessage {
    public var type: MetadataMessageType
    public var piece: Int
    public var totalSize: Int
    public var data: [UInt8]
}

/// Parses a ut_metadata message, splitting the trailing raw piece bytes from the
/// leading bencode dict.
public func parseMetadata(_ body: [UInt8]) throws -> MetadataMessage {
    let (v, n) = try bencodeDecodeN(body, 0)
    let d = try v.dictValue()
    var type = MetadataMessageType.request
    var piece = 0
    var totalSize = 0
    if let t = d.get("msg_type"), let ti = try? t.intValue() {
        type = MetadataMessageType(rawValue: Int(ti)) ?? .request
    }
    if let p = d.get("piece"), let pi = try? p.intValue() {
        piece = Int(pi)
    }
    if let ts = d.get("total_size"), let tsi = try? ts.intValue() {
        totalSize = Int(tsi)
    }
    return MetadataMessage(type: type, piece: piece, totalSize: totalSize, data: Array(body[n...]))
}

/// Reassembles the info dictionary from ut_metadata pieces and verifies it against
/// the expected info-hash.
public final class MetadataAssembler {
    private let totalSize: Int
    private var pieces: [Int: [UInt8]] = [:]

    /// Creates an assembler for a metadata of `totalSize` bytes.
    public init(totalSize: Int) {
        self.totalSize = totalSize
    }

    /// The number of 16 KiB pieces.
    public func pieceCount() -> Int {
        (totalSize + metadataPieceSize - 1) / metadataPieceSize
    }

    /// Stores a metadata piece.
    public func add(_ piece: Int, _ data: [UInt8]) {
        pieces[piece] = data
    }

    /// Reports whether every piece is present.
    public func isComplete() -> Bool { pieces.count == pieceCount() }

    /// Assembles the info dict and returns it if it matches `infoHash` (20 bytes).
    public func tryFinish(_ infoHash: [UInt8]) -> [UInt8]? {
        if !isComplete() { return nil }
        var out = [UInt8]()
        out.reserveCapacity(totalSize)
        for i in 0..<pieceCount() {
            out.append(contentsOf: pieces[i] ?? [])
        }
        if out.count != totalSize { return nil }
        if BTHash.sha1(out) != infoHash { return nil }
        return out
    }
}

// ── BEP-11 ut_pex ────────────────────────────────────────────────────────────

/// Builds a ut_pex message advertising added peers (compact form).
public func buildPexAdded(_ added: [PeerAddr]) -> [UInt8] {
    let d = BDict()
    try! d.add("added", .bytes(encodeCompactPeers(added)))
    return bencodeEncode(.dict(d))
}

/// Parses the "added" peers from a ut_pex message.
public func parsePexAdded(_ body: [UInt8]) throws -> [PeerAddr] {
    let v = try bencodeDecode(body)
    let d = try v.dictValue()
    if let a = d.get("added") {
        let b = try a.bytesValue()
        return try decodeCompactPeers(b)
    }
    return []
}

File: rust/src/bittorrent/extensions.rs
// SPDX-License-Identifier: MIT
//! BEP-10 extension protocol + BEP-9 ut_metadata + BEP-11 ut_pex.

use super::bencode::{self, Bencode};
use super::dht::{self, PeerAddr};
use sha1::{Digest, Sha1};
use std::collections::BTreeMap;

pub const EXTENDED_MESSAGE_ID: u8 = 20;
pub const EXTENSION_HANDSHAKE_ID: u8 = 0;
pub const METADATA_REQUEST: i64 = 0;
pub const METADATA_DATA: i64 = 1;
pub const METADATA_REJECT: i64 = 2;
pub const METADATA_PIECE_SIZE: usize = 16384;

pub fn wrap_extended(sub_id: u8, body: &[u8]) -> Vec<u8> {
    let mut out = vec![sub_id];
    out.extend_from_slice(body);
    out
}

pub fn split_extended(payload: &[u8]) -> Result<(u8, Vec<u8>), String> {
    if payload.is_empty() {
        return Err("empty extended payload".into());
    }
    Ok((payload[0], payload[1..].to_vec()))
}

pub fn build_extension_handshake(supported: &[(&str, i64)], metadata_size: i64) -> Vec<u8> {
    let mut m = BTreeMap::new();
    for (name, id) in supported {
        m.insert(name.as_bytes().to_vec(), Bencode::Int(*id));
    }
    let mut d = BTreeMap::new();
    d.insert(b"m".to_vec(), Bencode::Dict(m));
    if metadata_size > 0 {
        d.insert(b"metadata_size".to_vec(), Bencode::Int(metadata_size));
    }
    wrap_extended(EXTENSION_HANDSHAKE_ID, &bencode::encode(&Bencode::Dict(d)))
}

pub struct ExtensionHandshake {
    pub supported: BTreeMap<String, i64>,
    pub metadata_size: i64,
}

pub fn parse_extension_handshake(body: &[u8]) -> Result<ExtensionHandshake, String> {
    let d = bencode::decode(body)?;
    let dd = d.as_dict().ok_or("not a dict")?;
    let mut supported = BTreeMap::new();
    if let Some(m) = dd.get(b"m".as_slice()).and_then(|v| v.as_dict()) {
        for (k, v) in m {
            if let Some(id) = v.as_int() {
                supported.insert(String::from_utf8_lossy(k).into_owned(), id);
            }
        }
    }
    let metadata_size = dd.get(b"metadata_size".as_slice()).and_then(|v| v.as_int()).unwrap_or(0);
    Ok(ExtensionHandshake { supported, metadata_size })
}

pub fn build_metadata_request(piece: i64) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"msg_type".to_vec(), Bencode::Int(METADATA_REQUEST));
    d.insert(b"piece".to_vec(), Bencode::Int(piece));
    bencode::encode(&Bencode::Dict(d))
}

pub fn build_metadata_data(piece: i64, total_size: i64, data: &[u8]) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"msg_type".to_vec(), Bencode::Int(METADATA_DATA));
    d.insert(b"piece".to_vec(), Bencode::Int(piece));
    d.insert(b"total_size".to_vec(), Bencode::Int(total_size));
    let mut out = bencode::encode(&Bencode::Dict(d));
    out.extend_from_slice(data);
    out
}

pub fn build_metadata_reject(piece: i64) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"msg_type".to_vec(), Bencode::Int(METADATA_REJECT));
    d.insert(b"piece".to_vec(), Bencode::Int(piece));
    bencode::encode(&Bencode::Dict(d))
}

pub struct MetadataMessage {
    pub msg_type: i64,
    pub piece: i64,
    pub total_size: i64,
    pub data: Vec<u8>,
}

pub fn parse_metadata(body: &[u8]) -> Result<MetadataMessage, String> {
    let (v, n) = bencode::decode_n(body, 0)?;
    let d = v.as_dict().ok_or("not a dict")?;
    Ok(MetadataMessage {
        msg_type: d.get(b"msg_type".as_slice()).and_then(|x| x.as_int()).unwrap_or(0),
        piece: d.get(b"piece".as_slice()).and_then(|x| x.as_int()).unwrap_or(0),
        total_size: d.get(b"total_size".as_slice()).and_then(|x| x.as_int()).unwrap_or(0),
        data: body[n..].to_vec(),
    })
}

pub fn metadata_verify(pieces: &[Vec<u8>], total_size: usize, info_hash: &[u8; 20]) -> Option<Vec<u8>> {
    let out: Vec<u8> = pieces.concat();
    if out.len() != total_size {
        return None;
    }
    let mut h = Sha1::new();
    h.update(&out);
    if h.finalize().as_slice() == info_hash {
        Some(out)
    } else {
        None
    }
}

pub fn build_pex_added(added: &[PeerAddr]) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"added".to_vec(), Bencode::Bytes(dht::encode_compact_peers(added)));
    bencode::encode(&Bencode::Dict(d))
}

pub fn parse_pex_added(body: &[u8]) -> Result<Vec<PeerAddr>, String> {
    let d = bencode::decode(body)?;
    if let Some(added) = d.as_dict().and_then(|dd| dd.get(b"added".as_slice())).and_then(|v| v.as_bytes()) {
        return dht::decode_compact_peers(added);
    }
    Ok(vec![])
}
